### PR TITLE
Perf

### PR DESCRIPTION
**Pull request by blasten**

### CODE REVIEW

File: iron-list.html [hunk 1]
```diff
@@ -510,7 +510,7 @@
     /**
      * The max number of pages to render. One page is equivalent to the height of the list.
      */
-    _maxPages: 3,
+    _maxPages: 2,
```
Comment (blasten) [inline iron-list.html]: Tested on desktop and iPhone. This value provides good perf.

File: test/different-heights.html [hunk 2]
```diff
@@ -61,11 +61,10 @@
             list: list,
             contribution: 20,
             target: 100000,
+            delay: 1,
```
Comment (frankiefu) [inline test/different-heights.html]: Is the default for `delay` already set to `1`?

File: iron-list.html [hunk 3]
```diff
@@ -790,107 +786,84 @@
...
-        this._virtualStart = this._virtualStart + recycledTiles;
-        this._physicalStart = this._physicalStart + recycledTiles;
-        this._update(recycledTileSet, movingUp);
+        ith = this._physicalEnd;
+        lastIth = this._physicalStart;
+        offsetContent = this._physicalBottom - this._scrollBottom;
+      }
+
+      while (true) {
+        physicalItemHeight = this._getPhysicalSizeIncrement(ith);
```
Comment (frankiefu) [inline iron-list.html]: `physicalItemHeight` hasn't been declared yet.

File: test/different-heights.html (hunk 2 above)
Comment (blasten) [inline test/different-heights.html]: yeah, will remove

File: iron-list.html (hunk 3 above)
Comment (blasten) [inline iron-list.html]: good catch! fixed

**Comment by frankiefu** [comment]
LGTM

**Comment by Westbrook** [comment]
When ever I try to load `demo/grid.html` in this branch I get a handful of missing grid object like so:

screen shot 2016-09-21 at 3 26 08 pm

Is anyone else seeing this?

**Comment by blasten** [comment]
Good to know. I will check that!

**Comment by Westbrook** [comment]
When viewing `demo/grid.html`, after hitting the load threshold a number of time so that there is a large (500+) numbers of items loaded, drag scrolling to the top (hard edge of the scrollbar) of the list can find the user not at `items[0]`. Subsequent scrolling can correct the boundary, but once it has the number of child elements in `#items` begins to grows far beyond the size outlined by `_maxPages: 2`.

In my test (window size 1293 x 1031), an initial `childElementCount` of 42 for `#items` grows to 43 with a simplified downward scroll pattern that triggers 7 subsequent called to Flickr for images, leading to `items.length === 8000`. However, drag scrolling to the top causes the page to display like: 

screen shot 2016-09-26 at 12 02 09 pm

With `firstVisibleIndex === 352`. Further wheel based scroll then causes the scrollbar slider/handle to stutter until the actual top of the `item`s list is arrived at. From here, if you drag scroll to the bottom of the list, the `childElementCount` of `#items` will begin to grow, and ever subsequent drag scroll to full bottom will cause the `iron-list` to manage more physical DOM elements than the last.

**Comment by blasten** [comment]
@Westbrook would you mind testing one more time? After merging your patch that grid layout issue seems fixed.

**Comment by Westbrook** [comment]
I'm no longer seeing the erroneous edge loading.

However, I'm still seeing processing that bloats the number of physical items being tracked by the list. I may be abusing the interface a bit (I'm attempting to recreate something I have in a dev environment), but when you have a good number of items and you drag hard to the bottom and then hard back to the top (and then some) you can pause the renderer, and then it'll dive into an overlong pass in `_updateMetrics` during which time it'll jump into `_increasePool` erroneously. I've attempted to record my timeline, let me know if the attached file is helpful or if there is anything else I can provide to help find the a solution to this. It should be available here: https://www.wetransfer.com/downloads/c762b00d3f454ada3f52d0ce12a4407520160926212325/3593e287ef32387fc1263ae8ab91222820160926212325/b8c442 It was too big to attach inline.

**Comment by danbeam** [comment]
can we split the itemHeight/Size changes to a different PR?

**Comment by danbeam** [comment]
what's the hold up on this?  why can't we just do an element bump at https://github.com/PolymerElements/iron-list/pull/316/commits/742493427fb448dea77394c1ac3e819c0925bf5b ?

File: iron-list.html [hunk 4]
```diff
@@ -761,6 +770,20 @@
...
+        return list.grid ? node.getBoundingClientRect().width : node.offsetWidth;
+      } else {
+        return list.grid ? node.getBoundingClientRect().height : node.offsetHeight;
+      }
+    },
+
+    _itemSizeChanged: function(itemSize) {
+      var dataHost = this._getRootDataHost();
+      this._itemSizeFn = itemSize && (typeof itemSize == 'function' ? itemSize :
+        function() { return dataHost[itemSize].apply(dataHost, arguments); });
```
Comment (frankiefu) [inline iron-list.html]: Is `itemSize` a function?  What are you trying to do here?

Comment (blasten) [inline iron-list.html]: yep, it's a function. Users can bind to it in the same way they can bind to `dom-repeat`'s `sort` or `filter` property.

**Comment by blasten** [comment]
Closing this PR and following up in https://github.com/PolymerElements/iron-list/issues/331

**Pull request closed without merging**